### PR TITLE
Bank value plugin update

### PR DESCRIPTION
**Pull request by jplsek**
This is an update to TheLonelyDev's Bank Value Plugin (for #349). This ended up being mostly a rewrite of the plugin.

Any suggestions are welcome. Thanks.

Edit: here's a video:

ezgif-4-298ff5eca6
(Note: HA prices are disabled by default)

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/game/ItemManager.java [hunk 1]
```diff
@@ -191,16 +193,22 @@ public ItemComposition getItemComposition(int itemId)
...
 	 * @return
 	 */
-	public static String quantityToStackSize(int quantity)
+	public static String quantityToStackSize(long quantity)
 	{
-		if (quantity >= 10_000_000)
+		// Billion to show up after 10,000M (similar to 10,000K), since max stacks show 2147M
+		if (quantity >= 10_000_000_000L)
+		{
+			return quantity / 1_000_000_000L + "B";
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/game/ItemManager.java]: Don't add bills as then the quantity will be really inacurrate and we are following OSRS naming with this method, and it doesnt have bills, only K and M i believe

File: runelite-client/src/main/java/net/runelite/client/game/ItemManager.java [hunk 2]
```diff
@@ -43,6 +35,15 @@
 import net.runelite.http.api.item.ItemPrice;
 import net.runelite.http.api.item.SearchResult;
 
+import javax.annotation.Nullable;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/game/ItemManager.java]: Fix your imports: https://github.com/runelite/runelite/wiki/Code-Conventions

File: runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValueConfig.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import net.runelite.client.config.ConfigItem;
+
+@ConfigGroup(
+		keyName = "Bank Value",
+		name = "Bank Value plugin",
+		description = "Shows the value of your bank and/or current tab")
+public interface BankValueConfig extends Config
+{
+	@ConfigItem(
+			keyName = "showGE",
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValueConfig.java]: 2 tabs instead of 4

File: runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValueConfig.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+			name = "Show Grand Exchange price",
+			description = "Show grand exchange price total (GE)")
+	default boolean showGE()
+	{
+		return true;
+	}
+
+	@ConfigItem(
+			keyName = "showHA",
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValueConfig.java]: 2 tabs instead of 4

File: runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValuePlugin.java [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+import net.runelite.api.widgets.WidgetID;
+import net.runelite.api.widgets.WidgetInfo;
+import net.runelite.api.widgets.WidgetItem;
+import net.runelite.client.config.ConfigManager;
+import net.runelite.client.game.ItemManager;
+import net.runelite.client.plugins.Plugin;
+import net.runelite.client.plugins.PluginDescriptor;
+import net.runelite.http.api.item.ItemPrice;
+
+import javax.inject.Inject;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValuePlugin.java]: Fix your imports

**Comment by jplsek** [comment]
My most recent commit makes it work on any "starting" tab, instead of just the "all" tab. The tab that opens when opening the bank will give that tab the value. I disabled it for switching tabs because there is a problem with getDynamicItems in the BankItemQuery where it seems to "cache" the items in the bank when switching tabs. For example if you start on the "all" items tab, then switching to other tabs will be the same. If you start on tab "x", then switch to another tab, the value adds up every time. I don't know of a way to fix this.

So in summary:
Works:
- Calculate bank value on the "starting" tab

Does not work:
- Calculating after switching tabs

**Comment by jplsek** [comment]
Alright so now switching tabs work again.
The main thing I dislike is the sleep. It somewhat assumes that the bank opened within a second and all items are loaded. I could add back the interval check from the initial version if desired. The event checking is also not perfect, but I'm not sure on how to make it simpler.
Here's a screenshot:
image

File: runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValuePlugin.java [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+		}
+
+		ItemComposition itemComposition;
+
+		// Having this lock seems to not crash runescape. I'm not sure if it's because of this method call
+		// or if it's something else that this lock is slowing down. I believe it's because this is the only
+		// place that has a possibility of accessing runescape's code.
+		synchronized (rsItemLock)
+		{
+			itemComposition = itemManager.getItemComposition(item.getId());
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValuePlugin.java]: It is not safe to call this from any thread other than the game thread. I would gather all of these upfront.

File: runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValuePlugin.java [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+		{
+			bankThread.interrupt();
+		}
+	}
+
+	private void calculateBankWithThread()
+	{
+		bankActive = true;
+		log.debug("Bank value running");
+		bankThread = new Thread(this::calculateBank);
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValuePlugin.java]: There is a global ScheduledExecutorService you can inject and use instead of creating threads

File: runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValuePlugin.java [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+		}
+
+		WidgetItem[] widgetItems = new BankItemQuery().result(client);
+
+		if (widgetItems.length == 0)
+		{
+			return;
+		}
+
+		executor = Executors.newCachedThreadPool();
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValuePlugin.java]: I would probably make an api method on the web service that took in a list of itemids and mass-looked their prices up at once, so it could be in one request. And then maybe bypass item manager completely and use ItemClient.

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValuePlugin.java]: That would be definitely a lot better than creating new executor in separate thread and submitting bunch of tasks into it, that can't be correct no matter how I look at it.

File: runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValuePlugin.java [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+
+	class ItemPriceAndAlch extends ItemPrice
+	{
+		@Getter
+		@Setter
+		private int highAlchPrice;
+
+		ItemPriceAndAlch(int gePrice, int highAlchPrice)
+		{
+			setPrice(gePrice);
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValuePlugin.java]: Instead of doing this, add @Value annotation on ItemPriceAndAlch, then there would be no need for adding getter and setter annotation and making constructor

File: runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValuePlugin.java [hunk 10]
```diff
@@ -0,0 +1,381 @@
...
+	@Inject
+	private ItemManager itemManager;
+
+	private Thread bankThread = null;
+	private ExecutorService executor = null;
+
+	// Only meant to be used to tell if the player is in the bank (or during shutdown)
+	private boolean bankActive = false;
+
+	private String bankTitle = "The Bank of Runescape";
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValuePlugin.java]: This should be constant and not mutable variable, and remove the mutations of it.

**Comment by sethtroll** [comment]
maybe #881 deals with this sort of, as stuff like barrows items/imbued rings ect won't be calculated.

**Comment by Adam-** [comment]
I've added batch price lookup functionality server sided in 3ead452de9ece3740cc9c49f2a349b4aee1193fe.

Eg: https://api.runelite.net/runelite-1.3.0-SNAPSHOT/item/price?id=4151&id=4153

File: runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValuePlugin.java (hunk 10 above)
Comment (jplsek) [reply runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValuePlugin.java]: It's intentional. This variable gets set before the bank value gets put into the bank title. When changing the configs, the bank title gets reset by using that variable. This also goes for disabling the plugin while using the bank.

Comment (SomeoneWithAnInternetConnection) [reply runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValuePlugin.java]: I think deathbeam's comment applies here whether or not the mutability was intentional on your part. "The Bank of Runescape" is a string constant, so storing it in a mutable variable like this seems questionable at best

Comment (jplsek) [reply runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValuePlugin.java]: Okay, I see the point. It was partly done in case the bank title got called to be reset before the variable itself got updated. But in that case, the bank would already have the correct title, and would not need to be reset. I'll update that.

**Comment by jplsek** [comment]
Bulk price look up has been implemented. It also now uses game ticks to update the bank value. Now there doesn't need to be extra checks for config changes and tab switching/withdraw/deposit.

The current code is a little "busy." On every game tick, the bank gets recalculated (while using the cache). If we want it to be less busy, any suggestions?

Another possible change is when checking if the bank is active. It could be done in the game tick as well (checking if the bank title is shown). Would that be better than looking for a WidgetHiddenEvent? That would also remove the bankActive field.

During the rebase, should I remove the edits to ItemComposition, RSItemComposition, and Setting? They could be useful in the future.

File: runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValuePlugin.java [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+	}
+
+	@Override
+	protected void shutDown()
+	{
+		bankTitle.reset();
+	}
+
+	@Subscribe
+	public void onGameTick(GameTick event)
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValuePlugin.java]: Have you tried hooking to the widget (un)hidden event for the bank instead?

Comment (jplsek) [reply runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValuePlugin.java]: Yes, I tried it in the [previous commits](https://github.com/jplsek/runelite/blob/763adc1a6a7887b8f40994e155f097240aaea7b2/runelite-client/src/main/java/net/runelite/client/plugins/bankvalue/BankValuePlugin.java#L103) (if that's what your referring to). I removed it to avoid keeping track of a bankActive field.

Part of the problem is that the event gets called 1000's of times when opening the bank, and there doesn't seem to be a way to know when the bank is "done" loading. I tried to do it based on how many times the bank title widget gets updated, but it's a minimum of 4 and a maximum of how many tabs the player has. It falls apart when the user switches tabs/withdraw/deposit too quickly because only 2 events get fired, before calling 4 more. The code gets really ugly.

The WidgetHiddenChanged event can be put back, but calculating the bank using the async method that was added still needs to be done in the game tick (since the first call is usually null, so a second call needs to be made eventually). It would look the same as what was linked.

I would like to think that I'm just missing something really obvious, but I'm not sure.

**Comment by Adam-** [comment]
Earlier I changed this to use the item container for bank, but realized later maybe the widget query was intentional for per-tab totals?

**Comment by jplsek** [comment]
Possible. I had to modify [BankItemQuery](https://github.com/runelite/runelite/pull/824/files#diff-d5ff8e4963d1f1cc2d2f1786c0a2b0a2R64) because it would cache items and hide them when needed.
